pkg/gconfig/gconfig.go

```diff
@@ -0,0 +1,232 @@
+//
+// SSMDumper first pushes any updated secrets to ssm then returns the ssm paths to the secrets
+func (c Config) Dump(ctx context.Context, dumper Dumper) (map[string]string, error) {
+	if dumper == nil {
+		return nil, fmt.Errorf("cannot dump with nil dumper")
+	}
+	return dumper.Dump(ctx, c)
+}
+
+type Valuer interface {
+	Set(s string)
+	Get() string
+	String() string
+}
+type Field struct {
+	key      string
+	usage    string
+	value    Valuer
+	secret   bool
+	optional bool
+
+	// hasChanged is true if the Set() method has been called
+	hasChanged bool
+	// secretUpdated is true if the current value has been pushed to the secret backend e.g SSM
+	// This happens when Field.Dump(Dumper) is called with a secret dumper
+	secretUpdated bool
+	// secretPathPrefix defines the path that this secret should be written to.
+	// For example, in aws ssm, this is the secret path
+	secretPathPrefix string
+	// When a secret is read from file with the aws ssm loader, the path will be set here.
+	// If this is a newly created secret, when it is put in ssm, the path is saved here.
+	// this value is typically derived from the secretPathPrefix a suffix and a version number
+	secretPath string
+}
+
+func (s Field) HasChanged() bool {
+	return s.hasChanged
+}
+
+// Path returns the secret path
+// secrets loaded from config with the SSM Loader will have an secret path relevant to the loader type
+// secrets loaded from a test loader like JSONLoader or MapLoader will not have a path and this method will return an empty string
+func (s Field) SecretPath() string {
+	return s.secretPath
+}
+
+// IsSecret returns true if this Field is a secret
+func (s Field) IsSecret() bool {
+	return s.secret
+}
+
+// IsOptional returns true if this Field is optional
+func (s Field) IsOptional() bool {
+	return s.optional
+}
+
+// Key returns the key for this field
+func (s Field) Key() string {
+	return s.key
+}
+
+// Usage returns the usage string for this field
+func (s Field) Usage() string {
+	return s.usage
+}
+
+// Set the value of this string
+func (s *Field) Set(v string) error {
+	if s.value == nil {
+		return errors.New("cannot call Set on nil Valuer")
+	}
+	s.hasChanged = true
+	s.value.Set(v)
+	return nil
+}
+
+// Get returns the value if it is set, or an empty string if it is not set
+func (s *Field) Get() string {
+	if s.value == nil {
+		return ""
+	}
+	return s.value.Get()
+}
+
+// String calls the Valuer.String() method for this fields value.
+// If this field is a secret, then the response will be a redacted string.
+// Use Field.Get() to retrieve the raw value for the field
+func (s *Field) String() string {
+	if s.value == nil {
+		return ""
+	}
+	return s.value.String()
+}
+
+// SecretConfigValue value implements the Valuer interface, it should be used for secrets in configuration structs.
+//
+// It is configured to automatically redact the secret for common logging usecases like Zap, fmt.Println and json.Marshal
+type SecretConfigValue string
```

Comment:
same thinking as previous comment:

```suggestion
type SecretStringValue string
```